### PR TITLE
[Graph] add local scope function

### PR DESCRIPTION
**Pull request by jermainewang**
## Description
Using DGLGraph in NN module is quite painful because we always need to avoid (1) accidentally override existing feature data; (2) forget to pop the intermediate tensor. The PR added a `local_scope` function to `DGLGraph` that can solve this issue in one line:

Avoid accidentally overriding existing feature data. This is quite common when implementing a NN module:
```python
>>> def foo(g):
>>>     g = g.local_scope()
>>>     g.ndata['h'] = torch.ones((g.number_of_nodes(), 3))
>>>     return g.ndata['h']
>>>
>>> g = ... # some graph
>>> g.ndata['h'] = torch.zeros((g.number_of_nodes(), 3))
>>> newh = foo(g)  # get tensor of all ones
>>> print(g.ndata['h'])  # still get tensor of all zeros
```
Automatically garbage collect locally-defined tensors. Otherwise, user needs to call ``pop`` manually, which is error-prone:

```python
>>> def foo(g):
>>>     g = g.local_scope()
>>>     g.ndata['xxx'] = torch.ones((g.number_of_nodes(), 3))
>>>     return g.ndata['xxx']
>>>
>>> g = ... # some graph
>>> xxx = foo(g)
>>> print('xxx' in g.ndata)
False
```

@yzh119 please review. It should help your PR too.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

### CODE REVIEW

**Comment by VoVAllen** [comment]
How can I exit the local_scope?

**Comment by yzh119** [comment]
@VoVAllen `local_scope` returns a new graph object, it does not change the state of original graph, so there is no need to exit.
@jermainewang I wonder if we can implement it as a context manager?

**Comment by VoVAllen** [comment]
I think context manager is better if we would like to use the name "scope"

**Comment by jermainewang** [comment]
Hmm, it's quite different with context manager. Consider following example:
```python
def foo(G):
  with dgl.local_scope(G) as g:
    g.ndata['x'] = ...
  g.number_of_nodes()  # will this throw error?
  g.ndata['x']   # will this throw error?
```
In fact, the above code will throw no error. This is because `g` is not out-of-scope even if the `with` context has exited. The better implementation should be:
```python
def foo(G):
  with g.local_scope():
    # the __enter__ of the context manager replace g's frame with a new frame,
    # but also save the old frame
    g.ndata['x'] = ...
    # the __exit__ of the context manager recover the old frame.
```

Anyhow, I changed `local_scope` to `local_var` and implemented a new `local_scope` using context manager. I kept both syntax.

File: python/dgl/graph.py [hunk 1]
```diff
@@ -3344,3 +3345,110 @@ def to(self, ctx):
...
+        >>>     g.ndata['h'] = torch.ones((g.number_of_nodes(), 3))
+        >>>     return g.ndata['h']
+        >>>
+        >>> g = ... # some graph
+        >>> g.ndata['h'] = torch.zeros((g.number_of_nodes(), 3))
+        >>> newh = foo(g)  # get tensor of all ones
+        >>> print(g.ndata['h'])  # still get tensor of all zeros
+
+        Automatically garbage collect locally-defined tensors. Otherwise, user needs to
+        call ``pop`` manually, which is error-prone:
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/graph.py]: Done

File: python/dgl/graph.py [hunk 2]
```diff
@@ -3344,3 +3345,110 @@ def to(self, ctx):
...
+        >>>         g.ndata['h'] = torch.ones((g.number_of_nodes(), 3))
+        >>>         return g.ndata['h']
+        >>>
+        >>> g = ... # some graph
+        >>> g.ndata['h'] = torch.zeros((g.number_of_nodes(), 3))
+        >>> newh = foo(g)  # get tensor of all ones
+        >>> print(g.ndata['h'])  # still get tensor of all zeros
+
+        Automatically garbage collect locally-defined tensors. Otherwise, user needs to
+        call ``pop`` manually, which is error-prone:
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/graph.py]: Done

File: python/dgl/nn/mxnet/softmax.py [hunk 3]
```diff
@@ -30,56 +29,31 @@ def __init__(self, g):
         self.g = g
 
     def forward(self, score):
-        """
```
[the remark this answers is not in the document]
Comment (jermainewang) [reply python/dgl/nn/mxnet/softmax.py]: Done

**Comment by jermainewang** [comment]
@VoVAllen do you have any other comments?

**Comment by VoVAllen** [comment]
Looks good to me
